src/docs/binding-closures.md

```diff
@@ -0,0 +1,135 @@
+    public var context: AnyObject?
+}
+
+public struct Closure {
+    public var entryPoint: OpaquePointer
+    public var context: OpaquePointer?
+}
+```
+
+When invoking the closure, the self register needs to be set to contents of the context pointer.
+So essentially:
+```
+mov entryPoint[closoure], rax
+mov context[closure], r13
+jsr [rax]
+```
+or something similar.
+
+One interesting thing is that the Swift compiler writes closure with arguments for free variables as well as a forwarder.
+
+For example, if I have a function that returns a closure like this:
+
+```swift
+public func getSummer (a: Int) -> (Int) -> Int {
+    return { (b: Int) in
+        a + b
+    }
+}
+```
+Then the compiler will write this:
+```Swift
+private func getSummerImplementation(a: Int, b: Int) -> Int {
+    return a + b
+}
+```
+And a forwarder that looks like this:
+```
+move 10[r13], rsi // b goes into argument 2
+jmp _getSummerImplementation
+```
+
+# Runtime Differences
+
+Ideally, we would like to be able to pass C# delegates to Swift functions or store them into Swift types and have them call back into the right place.  The problem with this is that it would be essentially a reverse p/invoke so an arbitrary C# delegate is incompatible as is, but with support from the runtime, this should be less of an issue, but there are still some things that we would need to care about.
+
+If there is a case when we can't directly adapt a Swift closure into C#, there are some options that are available to us. In BTfS, which has no benefits from the runtime, this is done by converting Swift adapters into a more general form:
+
+Given a closure of the form `(arguments) -> return`, this can be converted into the form: `(UnsafeMutablePointer<return>, UnsafeMutablePointer<(arguments)>)->()` For example:
+```swift
+public func callsIntoCSharp (a: @escaping (Int, SomeStruct, Bool) -> SomeOtherStruct) {
+    let a_adapter = { (ret: UnsafeMutablePointer<SomeOtherStruct>, args: UnsafeMutablePointer<(Int, SomeStruct, Bool)>) in
+        let (i, ss, b) = args.pointee
+        ret.initialize(to: a(i, ss, b))
+    }
+    callToCSharpMethod (a_adapter)
+}
+```
+
+In this case, the original closure gets captured by `a_adapter`. Before calling the adapter, C# obviously has to create the argument tuple, space for the return value (allocated but not initialized) and can call it because effectively the delegate signature has become `delegate void csAdapter (IntPtr ret, IntPtr args)`. A similar process is used to get C# closures into Swift.
+
+One obvious problem here is that this only works with escaping closures. If rewritten with a non escaping closure, the act of capturing the original closure in the new one is flagged by the Swift compiler as an error. Fortunately, Swift has a workaround to this via the function [`withoutActuallyEscaping`](https://developer.apple.com/documentation/swift/withoutactuallyescaping(_:do:)/), for which the previous code can be rewritten as this:
```

Comment:
```suggestion
One obvious problem here is that this only works with escaping closures. If rewritten with a non-escaping closure, the act of capturing the original closure in the new one is flagged by the Swift compiler as an error. Fortunately, Swift has a workaround to this via the function [`withoutActuallyEscaping`](https://developer.apple.com/documentation/swift/withoutactuallyescaping(_:do:)/), for which the previous code can be rewritten as this:
```

Reply:
@matouskozak The data context object always shows up in the self register. From a consistency standpoint, it makes sense as you can look at a closure with captured data as being equivalent to an instance method on an anonymous type (class if it's escaping, value if it's non-escaping). From this point of view, the context is 100% equivalent to `self`.